databricks/koalas/groupby.py

```diff
@@ -1401,6 +1401,67 @@ def ffill(self, limit=None):
 
     pad = ffill
 
+    def head(self, n=5):
+        """
+        Return first n rows of each group.
+
+        Returns
+        -------
+        DataFrame or Series
+
+        Examples
+        --------
+        >>> df = ks.DataFrame({'a': [1, 1, 1, 1, 2, 2, 2, 3, 3, 3],
+        ...                    'b': [2, 3, 1, 4, 6, 9, 8, 10, 7, 5],
+        ...                    'c': [3, 5, 2, 5, 1, 2, 6, 4, 3, 6]},
+        ...                   columns=['a', 'b', 'c'],
+        ...                   index=[7, 2, 4, 1, 3, 4, 9, 10, 5, 6])
+        >>> df
+            a   b  c
+        7   1   2  3
+        2   1   3  5
+        4   1   1  2
+        1   1   4  5
+        3   2   6  1
+        4   2   9  2
+        9   2   8  6
+        10  3  10  4
+        5   3   7  3
+        6   3   5  6
+
+        >>> df.groupby('a').head(2)
+            a   b  c
+        7   1   2  3
+        2   1   3  5
+        10  3  10  4
+        5   3   7  3
+        3   2   6  1
+        4   2   9  2
+
+        >>> df.groupby('a')['b'].head(2)
+        7      2
+        2      3
+        10    10
+        5      7
+        3      6
+        4      9
+        Name: b, dtype: int64
+        """
+        groupkeys = self._groupkeys
+        tmp_col = '__order__'
+        sdf = self._kdf._sdf.withColumn(tmp_col, monotonically_increasing_id())
+        name = self._agg_columns[0]._internal.data_columns[0]
+        window = Window.partitionBy([s._scol for s in groupkeys]).orderBy(F.col(tmp_col))
+        sdf = sdf.withColumn(tmp_col, F.row_number().over(window)).filter(F.col(tmp_col) <= n)
```

Comment:
`monotonically_increasing_id()` here wouldn't work if there are multiple partitions because `monotonically_increasing_id()` isn't sequential:

```
scala> spark.range(100).repartition(200).select(monotonically_increasing_id()).show()
+-----------------------------+
|monotonically_increasing_id()|
+-----------------------------+
|                 103079215104|
|                 111669149696|
|                 120259084288|
...
```

Reply:
I think we can just use `row_number`

Reply:
doesn't function `head` pick the top N values without sort the order?

if we use `row_number`, i think we have to use it with `orderBy`,

then, the existing order is sorted so that we couldn't get the proper result?

For example, we have dataframe like this.

```python
>>> df = ks.DataFrame({'a': [1, 1, 1, 1, 2, 2, 2, 3, 3, 3],
...                    'b': [2, 3, 1, 4, 6, 9, 8, 10, 7, 5],
...                    'c': [3, 5, 2, 5, 1, 2, 6, 4, 3, 6]},
...                   columns=['a', 'b', 'c'],
...                   index=[7, 2, 4, 1, 3, 4, 9, 10, 5, 6])

>>> df
    a   b  c
7   1   2  3
2   1   3  5
4   1   1  2
1   1   4  5
3   2   6  1
4   2   9  2
9   2   8  6
10  3  10  4
5   3   7  3
6   3   5  6
```

we expect the result of `df.groupby('a').head(2)` like below.

```python
>>> df.groupby('a').head(2)
    a   b  c
7   1   2  3
2   1   3  5
3   2   6  1
4   2   9  2
10  3  10  4
5   3   7  3
```

but if we use `row_number`, maybe the result will be shown like below.

```python
   a  b  c
4  1  1  2
7  1  2  3
6  3  5  6
5  3  7  3
3  2  6  1
9  2  8  6
```

is this result seems okay??

Reply:
It's already sorted by `monotonically_increasing_id` so using `row_number` won't make much differences. We cannot use `F.col(tmp_col) <= n` as is anyway because `tmp_col` will be bigger than `n`.